AX: iOS should let native focus handle presenting date time picker

hoffmanjoshua opened the pull request:
#### f087f40788fa965aa56aee010b4a4eeac7755b18
<pre>
AX: iOS should let native focus handle presenting date time picker
<a href="https://bugs.webkit.org/show_bug.cgi?id=304364">https://bugs.webkit.org/show_bug.cgi?id=304364</a>
<a href="https://rdar.apple.com/165526147">rdar://165526147</a>

Reviewed by Tyler Wilcock and Chris Fleizach.

In 301528@main, date and datetime inputs started being treated as action elements. This
is fine for platforms where WebKit handles the creation and interaction of the date time
picker. But, on iOS, `showPicker` doesn&apos;t do anything. It instead relies on the peripheral
input system, which manages the presentation of the picker.

We should use the presence of segments to determine how we should handle the press, as exposed
through hasCustomFocusLogic on the date input. This will be different between the platforms,
as iOS doesn&apos;t have segments and doesn&apos;t manage it&apos;s own focus.

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::isDateFieldWithStandardFocus):
(WebCore::nodeActionElement):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: <a href="https://commits.webkit.org/304700@main">https://commits.webkit.org/304700@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31b5e592743c9b2723a97b6e8d2c969026c113b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136295 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8652 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144006 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d0cf8691-1443-4699-8efa-66af5eb5e630) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9331 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8496 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104216 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9c236ccc-eba2-477a-aac5-ec8c467ef3f7) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139240 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6792 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122131 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85049 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5928089d-27d8-4e03-be97-8e906c112253) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6445 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4106 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4598 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115736 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146750 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8334 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40900 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112555 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8351 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7004 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112899 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28662 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6376 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118436 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8382 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36493 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8100 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71941 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8322 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8174 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1523,7 +1523,11 @@ static RefPtr<Element> nodeActionElement(Node& node)
     auto elementName = WebCore::elementName(node);
     if (RefPtr input = dynamicDowncast<HTMLInputElement>(node)) {
-        if (!input->isDisabledFormControl() && (input->isRadioButton() || input->isCheckbox() || input->isTextButton() || input->isFileUpload() || input->isImageButton() || input->isTextField() || input->isDateField() || input->isDateTimeLocalField()))
+        if (!input->isDisabledFormControl() && (input->isRadioButton() || input->isCheckbox() || input->isTextButton() || input->isFileUpload() || input->isImageButton() || input->isTextField()
```
I think this would be cleaner if you made a static helper function, like isActionInputElement, and put the platform guard in there.

I also wonder if there's some way to check the renderer and see if it's actually a real date time renderer or just a button that pops up a native picker. That seems like it'd be the more proper check.

hoffmanjoshua replied:
My new patch addresses this—we check the `hasCustomFocusLogic` state of the input.

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1421,12 +1421,14 @@ bool AccessibilityObject::press()
     if (dispatchedEvent)
         return true;
 
+#if !PLATFORM(IOS_FAMILY)
```
I think it'd be cleaner to make a similar helper function (or share one with the one above) that returns whether an input element is a webkit date time renderer or not. Ideally that would check the renderer, but worst case, bury the platform #define in there.

hoffmanjoshua replied:
Removed this platform guard entirely in latest patch

minorninth commented on Source/WebCore/html/BaseDateAndTimeInputType.cpp:
```diff
@@ -339,6 +339,7 @@ void BaseDateAndTimeInputType::handleDOMActivateEvent(Event& event)
     showPicker();
 }
 
+#if !PLATFORM(IOS_FAMILY)
```
I'd avoid platform guards in this file. We should just avoid calling this method on iOS in accessibility-specific logic.

hoffmanjoshua replied:
Removed this

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1519,11 +1519,16 @@ bool AccessibilityNodeObject::isHiddenUntilFoundContainer() const
     auto elementName = WebCore::elementName(node);
     if (RefPtr input = dynamicDowncast<HTMLInputElement>(node)) {
-        if (!input->isDisabledFormControl() && (input->isRadioButton() || input->isCheckbox() || input->isTextButton() || input->isFileUpload() || input->isImageButton() || input->isTextField() || input->isDateField() || input->isDateTimeLocalField()))
+        if (!input->isDisabledFormControl() && (input->isRadioButton() || input->isCheckbox() || input->isTextButton() || input->isFileUpload() || input->isImageButton() || input->isTextField() || isDateFieldWithStandardFocus(*input)))
```
Can we add a comment explaining why a date field needs "standard focus" to be an action element?

hoffmanjoshua replied:
Done!

fleizach reviewed (approved):
Great fix

hoffmanjoshua committed:
AX: iOS should let native focus handle presenting date time picker
https://bugs.webkit.org/show_bug.cgi?id=304364
rdar://165526147

Reviewed by Tyler Wilcock and Chris Fleizach.

In 301528@main, date and datetime inputs started being treated as action elements. This
is fine for platforms where WebKit handles the creation and interaction of the date time
picker. But, on iOS, `showPicker` doesn't do anything. It instead relies on the peripheral
input system, which manages the presentation of the picker.

We should use the presence of segments to determine how we should handle the press, as exposed
through hasCustomFocusLogic on the date input. This will be different between the platforms,
as iOS doesn't have segments and doesn't manage it's own focus.

* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::isDateFieldWithStandardFocus):
(WebCore::nodeActionElement):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: https://commits.webkit.org/304700@main
